### PR TITLE
#381 Move scan formatters to a new "formattedcode" module

### PR DESCRIPTION
**Pull request by pombredanne**
* this is a first step towards #381
 * code and tests have been split from the scancode module
 * Also restore the ability to run ScanCode as a library by vendoring
   the NOTICE file. This was a regression introduced in #424 and
   tracked as #468
   

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

### CODE REVIEW

**Comment by pombredanne** [comment]
@sschuberth Input welcome. This is a first toward creating output plugins.

File: src/formattedcode/saver.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        # Remove duplicate licenses from the list for the package.
+        unique_licenses = set(doc.package.licenses_from_files)
+        if len(doc.package.licenses_from_files) == 0:
+            if all_files_have_no_license:
+                doc.package.licenses_from_files = [SPDXNone()]
+            else:
+                doc.package.licenses_from_files = [NoAssert()]
+        else:
+            # List license identifiers alphabetically for the package.
+            doc.package.licenses_from_files = sorted(unique_licenses, key = lambda x : x.identifier)
```
Comment (sschuberth) [inline src/formattedcode/saver.py]: Nit: Superfluous space before `:`?

Comment (pombredanne) [inline src/formattedcode/saver.py]: good catch.

**Comment by pombredanne** [comment]
@sschuberth you wrote:
> Looks good to me, except that out of personal preference I'd use the term "writer(s)" instead of "saver(s)" in the file names, and I would also call the methods "write_<format_name>".

Let me rename the module and the one function

**Comment by pombredanne** [comment]
done.

**Comment by pombredanne** [comment]
I am repspinning the failed Windows build, though this is likely a transient failure.

**Comment by pombredanne** [comment]
@sschuberth Thanks for the review and feedback. Merging.

**Comment by sschuberth** [comment]
> Also restore the ability to run ScanCode as a library by vendoring
the NOTICE file.

FYI, your fix for this regression introduced another regression :-) Now, we cannot run `scancode` from another directory than the scancode directory itself anymore. I.e. running

    scancode-toolkit/scancode -f json scancode-toolkit/AUTHORS.rst

results in

      File "/home/seschube/Development/HERE/oss/tools/license-scanner-framework/scanners/scancode-toolkit/src/scancode/cli.py", line 82, in <module>
        notice_text = open('NOTICE').read()
    IOError: [Errno 2] No such file or directory: u'NOTICE'

**Comment by pombredanne** [comment]
@sschuberth Sorry for that! let me get to the bottom of this right away!

**Comment by sschuberth** [comment]
No worries, I've applied a hotfix for this on our side. I was wondering if we could look at `os.path.dirname(__file__)` to decide whether we're running inside a wheel or not, and use the correct path to `NOTICE` accordingly. But maybe there's a better way to check if we're running a wheel...

**Comment by pombredanne** [comment]
@sschuberth This is eventually the correct way. Let's track this in #506

**Pull request merged**